### PR TITLE
Support for @ConfigMapping

### LINKED ISSUE #413: Support for @ConfigMapping
@ConfigProperties from Quarkus seems to be deprecated and replaced with @ConfigMapping https://github.com/quarkusio/quarkus/blob/2e5a2c7a039ecdb1e0997bf6197f2b2edb1485e2/docs/src/main/asciidoc/config-mappings.adoc#2-configproperties-deprecated

See https://quarkus.io/guides/config-mappings

### PR DESCRIPTION
**Pull request by angelozerr**
Support for @ConfigMapping

Fixes #413

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
I noticed that there are some duplicate property with the following use case:

```java
package org.acme.namingstrategy;

import io.smallrye.config.ConfigMapping;

@ConfigMapping(prefix = "server.kebab", namingStrategy = ConfigMapping.NamingStrategy.KEBAB_CASE)
public interface ServerKebabCaseNamingStrategy {
  
	String theHost();

	int thePort();
}
```

You will see in the completion 2 properties with `server.kebab.the-host`, although if you write:

```java
import static io.smallrye.config.ConfigMapping.NamingStrategy.KEBAB_CASE;

@ConfigMapping(prefix = "server.kebab", namingStrategy = KEBAB_CASE)
```

It works good (no duplicate properties).

The reason of this problem is that search engine returns 2 references for ServerKebabCaseNamingStrategy class:

 * reference for @ConfigMapping(prefix,...  
 *  reference for  ConfigMapping.NamingStrategy.KEBAB_CASE

I have not find a solution to customize that with Java SearchPattern:

```
String annotationName = "ConfigMapping";
SearchPattern.createPattern(annotationName, IJavaSearchConstants.ANNOTATION_TYPE,
				IJavaSearchConstants.ANNOTATION_TYPE_REFERENCE, SearchPattern.R_EXACT_MATCH);
```

ANNOTATION_TYPE_REFERENCE returns all references. I would like to create an issue for that and work on it in an another issue.

**Comment by angelozerr** [comment]
This PR doesn't support:

 * https://quarkus.io/guides/config-mappings#collections
 * https://quarkus.io/guides/config-mappings#maps

It requires to do some improvement on LSP4MP, that's why I would like to manage those usecases in another issue.

**Comment by rgrunber** [comment]
The change seems to be working pretty well! I've tested with `prefix`, the various `namingStrategy` values, `WithName`, `WithParentName`, and `WithDefault`. All seem to be working as expected.

I would just have a final look at QuarkusConfigMappingProvider.java and then I think we can prepare to merge.

**Comment by rgrunber** [comment]
I can confirm using just `@ConfigMapping` (no prefix member) results in :

```
at io.smallrye.config.ConfigMappingProvider.mapConfiguration(ConfigMappingProvider.java:818)
        at io.smallrye.config.ConfigMappingProvider.mapConfiguration(ConfigMappingProvider.java:777)
        at io.smallrye.config.ConfigMappings.mapConfiguration(ConfigMappings.java:54)
        at io.smallrye.config.ConfigMappings.registerConfigMappings(ConfigMappings.java:36)
        at io.quarkus.arc.runtime.ConfigRecorder.registerConfigMappings(ConfigRecorder.java:50)
        at io.quarkus.deployment.steps.ConfigBuildStep$registerConfigClasses-2116437784.deploy_0(ConfigBuildStep$registerConfigClasses-2116437784.zig:132)
        at io.quarkus.deployment.steps.ConfigBuildStep$registerConfigClasses-2116437784.deploy(ConfigBuildStep$registerConfigClasses-2116437784.zig:40)
        at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:623)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:101)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:66)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:42)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:119)
        at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:98)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

although we certainly support the empty-form in our code. I get the impression this should probably be filed against Quarkus because there needs to be a better error message if this is truly not supported. We can also address it in future changes taht adjust validation.

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/core/properties/QuarkusConfigMappingProvider.java [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+			IProgressMonitor monitor) throws JavaModelException {
+		if (typesAlreadyProcessed.contains(configMappingType)) {
+			return;
+		}
+		typesAlreadyProcessed.add(configMappingType);
+		IJavaElement[] elements = configMappingType.getChildren();
+		// Loop for each methods
+		for (IJavaElement child : elements) {
+			if (child.getElementType() == IJavaElement.METHOD) {
+				IMethod method = (IMethod) child;
```
Comment (rgrunber) [inline quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/core/properties/QuarkusConfigMappingProvider.java:134]: You asked if we could get the information about generics.

Supposed we have `List<String>`, then `method.getReturnType()` gives `QArrayList<QString;>;` . It's in there, but I haven't found a utility method to get the generic element (`QString`) but I'm pretty sure it should be possible.

**Comment by rgrunber** [comment]
I did find something we should probably try to address. Add a method to the interface like `ArrayList<String> getOptions ();` . We don't support ArrayList, but going into the `application.properties` and triggering completion results in a huge processing time that ultimately ends with :


<details>
<summary>stacktrace</summary>

```
[Error - 1:49:33 PM] Nov 5, 2021, 1:49:33 PM An internal error occurred during: "MicroProfile properties collector".
null
java.lang.StackOverflowError
	at java.base/java.lang.StackStreamFactory$AbstractStackWalker.fetchStackFrames(StackStreamFactory.java:384)
	at java.base/java.lang.StackStreamFactory$AbstractStackWalker.getNextBatch(StackStreamFactory.java:322)
	at java.base/java.lang.StackStreamFactory$AbstractStackWalker.peekFrame(StackStreamFactory.java:263)
	at java.base/java.lang.StackStreamFactory$AbstractStackWalker.hasNext(StackStreamFactory.java:351)
	at java.base/java.lang.StackStreamFactory$StackFrameTraverser.tryAdvance(StackStreamFactory.java:593)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at java.logging/java.util.logging.LogRecord$CallerFinder.lambda$get$1(LogRecord.java:709)
	at java.base/java.lang.StackStreamFactory$StackFrameTraverser.consumeFrames(StackStreamFactory.java:534)
	at java.base/java.lang.StackStreamFactory$AbstractStackWalker.doStackWalk(StackStreamFactory.java:306)
	at java.base/java.lang.StackStreamFactory$AbstractStackWalker.callStackWalk(Native Method)
	at java.base/java.lang.StackStreamFactory$AbstractStackWalker.beginStackWalk(StackStreamFactory.java:370)
	at java.base/java.lang.StackStreamFactory$AbstractStackWalker.walk(StackStreamFactory.java:243)
	at java.base/java.lang.StackWalker.walk(StackWalker.java:498)
	at java.logging/java.util.logging.LogRecord$CallerFinder.get(LogRecord.java:709)
	at java.logging/java.util.logging.LogRecord.inferCaller(LogRecord.java:683)
	at java.logging/java.util.logging.LogRecord.getSourceClassName(LogRecord.java:352)
	at java.logging/java.util.logging.SimpleFormatter.format(SimpleFormatter.java:160)
	at java.logging/java.util.logging.StreamHandler.publish(StreamHandler.java:199)
	at java.logging/java.util.logging.ConsoleHandler.publish(ConsoleHandler.java:95)
	at java.logging/java.util.logging.Logger.log(Logger.java:979)
	at java.logging/java.util.logging.Logger.doLog(Logger.java:1006)
	at java.logging/java.util.logging.Logger.log(Logger.java:1029)
	at com.redhat.microprofile.jdt.internal.quarkus.core.properties.QuarkusConfigMappingProvider.populateConfigObject(QuarkusConfigMappingProvider.java:139)
	at com.redhat.microprofile.jdt.internal.quarkus.core.properties.QuarkusConfigMappingProvider.populateConfigObject(QuarkusConfigMappingProvider.java:184)
	at com.redhat.microprofile.jdt.internal.quarkus.core.properties.QuarkusConfigMappingProvider.populateConfigObject(QuarkusConfigMappingProvider.java:184)
	at com.redhat.microprofile.jdt.internal.quarkus.core.properties.QuarkusConfigMappingProvider.populateConfigObject(QuarkusConfigMappingProvider.java:184)
	at com.redhat.microprofile.jdt.internal.quarkus.core.properties.QuarkusConfigMappingProvider.populateConfigObject(QuarkusConfigMappingProvider.java:184)
	at com.redhat.microprofile.jdt.internal.quarkus.core.properties.QuarkusConfigMappingProvider.populateConfigObject(QuarkusConfigMappingProvider.java:184)
	at com.redhat.microprofile.jdt.internal.quarkus.core.properties.QuarkusConfigMappingProvider.populateConfigObject(QuarkusConfigMappingProvider.java:184)
```
</details>

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/core/properties/QuarkusConfigMappingProvider.java (hunk 1 above)
Comment (angelozerr) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/core/properties/QuarkusConfigMappingProvider.java:134]: Thanksforthe feedback!

**Comment by rgrunber** [comment]
I'm fine with not supporting the "Converter" case for now because as you said, we don't currently have a way to check if Quarkus has a converter for the given class. My only concern for now is avoiding things like stack overflows or any behaviour that makes editing unusable.

**Comment by rgrunber** [comment]
Once the tests are adjusted, feel free to merge. I think we can address any remaining issues in the subsequent PRs that will handle more strict validation.

**Pull request merged**